cloud-controller-manager/do/resources_test.go

```diff
@@ -37,7 +39,352 @@ import (
+			foundLB: nil,
+			found:   false,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			resources := newResources()
+			resources.UpdateLoadBalancers(test.lbs)
+
+			lb, found := resources.LoadBalancerByID(test.findID)
+			if want, got := test.found, found; want != got {
+				t.Errorf("incorrect found\nwant: %#v\n got: %#v", want, got)
+			}
+			if want, got := test.foundLB, lb; !reflect.DeepEqual(want, got) {
+				t.Errorf("incorrect lb\nwant: %#v\n got: %#v", want, got)
+			}
+		})
+	}
+}
+
+func TestResources_LoadBalancers(t *testing.T) {
+	lbs := []*godo.LoadBalancer{{ID: "1"}, {ID: "2"}}
+	resources := &resources{
+		loadBalancerIDMap: map[string]*godo.LoadBalancer{
+			lbs[0].ID: lbs[0],
+			lbs[1].ID: lbs[1],
+		},
+	}
+
+	foundLBs := resources.LoadBalancers()
+	// order found lbs by id so we can compare
+	sort.Slice(foundLBs, func(a, b int) bool { return foundLBs[a].ID < foundLBs[b].ID })
+	if want, got := lbs, foundLBs; !reflect.DeepEqual(want, got) {
+		t.Errorf("incorrect lbs\nwant: %#v\n got: %#v", want, got)
+	}
+}
+
+type recordingSyncer struct {
+	*tickerSyncer
+
+	synced map[string]int
+	mutex  sync.Mutex
+}
+
+func newRecordingSyncer() *recordingSyncer {
+	return &recordingSyncer{
+		tickerSyncer: &tickerSyncer{},
+		synced:       make(map[string]int),
+	}
+}
+
+func (s *recordingSyncer) Sync(name string, period time.Duration, stopCh <-chan struct{}, fn func() error) {
+	recordingFn := func() error {
+		s.mutex.Lock()
+		defer s.mutex.Unlock()
+
+		count, _ := s.synced[name]
+		s.synced[name] = count + 1
+
+		return fn()
+	}
+
+	s.tickerSyncer.Sync(name, period, stopCh, recordingFn)
+}
+
+var (
+	clusterID    = "0caf4c4e-e835-4a05-9ee8-5726bb66ab07"
+	clusterIDTag = buildK8sTag(clusterID)
+)
+
+func TestResourcesController_Run(t *testing.T) {
+	fakeResources := newResources()
+	kclient := fake.NewSimpleClientset()
+	inf := informers.NewSharedInformerFactory(kclient, 0)
+	gclient := &godo.Client{
+		Droplets: &fakeDropletService{
+			listFunc: func(ctx context.Context, opt *godo.ListOptions) ([]godo.Droplet, *godo.Response, error) {
+				return []godo.Droplet{{ID: 2, Name: "two"}}, newFakeOKResponse(), nil
+			},
+		},
+		LoadBalancers: &fakeLBService{
+			listFn: func(ctx context.Context, opt *godo.ListOptions) ([]godo.LoadBalancer, *godo.Response, error) {
+				return []godo.LoadBalancer{{ID: "2", Name: "two"}}, newFakeOKResponse(), nil
+			},
+		},
+	}
+
+	res := NewResourcesController(clusterID, fakeResources, inf.Core().V1().Services(), kclient, gclient)
+	syncer := newRecordingSyncer()
+	res.syncer = syncer
+
+	stop := make(chan struct{})
+	res.Run(stop)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				stop <- struct{}{}
+			default:
+				syncer.mutex.Lock()
+				if syncer.synced["resources syncer"] > 0 && syncer.synced["tags syncer"] > 0 {
```

Comment:
This is going to busy-spin for up to 3 seconds while locking/unlocking continuously. I think a more elegant solution could be the following:

- make `newRecordingSyncer` accept the stop channel (and add a field to `recordingSyncer` for storing it)
- let `recorderSyncer`'s `recordingFn` close the stop signal as soon as it has been called once for each name
- reduce the `go func()` in the test to a simple `select` between the stop signal firing and a custom timeout defining an upper timeout for the test (like the current 3 seconds)
- call `t.Errorf` straight when the timeout has been hit and include the number of invocations per name

I believe this would remove the busy spinning, the locking, and simplify the test overall.

WDYT?